src/Toml/Codec/Di.hs

```diff
@@ -0,0 +1,210 @@
+{- |
+Copyright: (c) 2018-2020 Kowainik
+SPDX-License-Identifier: MPL-2.0
+Maintainer: Kowainik <xrom.xkov@gmail.com>
+
+Forward and backward mapping functions and combinators (similar to profunctors).
+-}
+
+module Toml.Codec.Di
+    ( dimap
+    , dioptional
+    , diwrap
+    , dimatch
+    , (<!>)
+    , (.=)
+
+    ) where
+
+import Control.Applicative (Alternative (..), optional)
+import Data.Coerce (Coercible, coerce)
+
+import Toml.Codec.Types (Codec (..), TomlCodec)
+
+
+{- | This is an instance of @Profunctor@ for 'Codec'. But since there's no
+@Profunctor@ type class in @base@ or package with no dependencies (and we don't
+want to bring extra dependencies) this instance is implemented as a single
+top-level function.
+
+Useful when you want to parse @newtype@s. For example, if you had data type like
+this:
+
+@
+__data__ Example = Example
+    { foo :: Bool
+    , bar :: Text
+    }
+@
+
+Bidirectional TOML converter for this type will look like this:
+
+@
+exampleCodec :: TomlCodec Example
+exampleCodec = Example
+    \<$\> Toml.bool "foo" '.=' foo
+    \<*\> Toml.text "bar" '.=' bar
+@
+
+Now if you change your type in the following way:
+
+@
+__newtype__ Email = Email { unEmail :: Text }
+
+__data__ Example = Example
+    { foo :: Bool
+    , bar :: Email
+    }
+@
+
+you need to patch your TOML codec like this:
+
+@
+exampleCodec :: TomlCodec Example
+exampleCodec = Example
+    \<$\> Toml.bool "foo" '.=' foo
+    \<*\> 'dimap' unEmail Email (Toml.text "bar") '.=' bar
+@
+-}
+dimap
+    :: (inC -> inD)    -- ^ Mapper for consumer
+    -> (outA -> outB)  -- ^ Mapper for producer
+    -> Codec inD outA  -- ^ Source 'Codec' object
+    -> Codec inC outB  -- ^ Target 'Codec' object
```

Comment:
While working with `tomland`, I noticed that it's confusing to have these functions work with `Codec` when in reality they work with `TomlCodec` all the time. I used the wrong functions and got scary error messages. So I would propose to specialize types of all functions in here for `TomlCodec` to also have a more readable type signatures.

```suggestion
dimap
    :: (b -> a)    -- ^ Mapper for consumer
    -> (a -> b)  -- ^ Mapper for producer
    -> TomlCodec a  -- ^ Source 'Codec' object
    -> TomlCodec b  -- ^ Target 'Codec' object
```